README.md
# Electrify [![windows](https://img.shields.io/appveyor/ci/arboleya/electrify.svg?label=windows)](https://ci.appveyor.com/project/arboleya/electrify) [![travis](https://img.shields.io/travis/arboleya/electrify/master.svg?label=osx/linux)](https://travis-ci.org/arboleya/electrify) [![coverage](https://img.shields.io/codeclimate/coverage/github/arboleya/electrify.svg)](https://codeclimate.com/github/arboleya/electrify/coverage) [![join the chat at https://gitter.im/arboleya/electrify](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/arboleya/electrify)

Easily package your Meteor apps with Electron, and *butter*.

## TL;DR

````shell
npm install -g electrify
cd /your/meteor/app
electrify
````
## Compatibility

Works on all Meteor's supported [platforms](https://github.com/meteor/meteor/wiki/Supported-Platforms).

## Help

````bash
$ electrify -h

  Usage: electrify [command] [options]

  Commands:

    run       (default) start meteor app within electrify context
    bundle    bundle meteor app at `.electrify` dir
    package   bundle and package app to `--output` dir

  Options:

    -h, --help             output usage information
    -V, --version          output the version number
    -i, --input    <path>  meteor app dir       | default = .
    -o, --output   <path>  output dir           | default = .electrify/.dist
    -s, --settings <path>  meteor settings file | default = null (optional)

  Examples:

    # cd into meteor dir first
    cd /your/meteor/app

    electrify
    electrify run
    electrify package
    electrify package -o /dist/dir
    electrify package -o /dist/dir -s file.json
    electrify package -i /app/dir -o /dist/dir -s dev.json
    electrify package -- <electron-packager-options>

    # more info about electron packager options:
    # ~> https://www.npmjs.com/package/electron-packager
````

## Installation

````shell
npm install -g electrify
````

> For invoking Electron methods from Meteor, you'll also need to install the
> `arboleya:electrify` meteor package.
> For more info check [Meteor x Electron integration](#meteor-x-electron-integration).

## Running app

````shell
cd /your/meteor/app
electrify
````

## Packaging

````shell
cd /your/meteor/app
electrify package
````

The packaging process is done under the hood using `electron-packager`
npm package. The following variables are automatically set:

  * `--out` -- *comes from cli option [-o, --out]*
  * `--arch` -- *comes from system [current arch]*
  * `--platform` -- *comes from system [current platform]*
  * `--version` -- *comes from .electrify/package.json [current app version]*

You can overwrite these default values and also set others by passing custom
arguments directly to `electron-packager` after `--`, i.e:

````shell
cd /your/meteor/app
electrify package -- --icon=/folder/x/img/icon.png --version=x.y.z
````

All the available options for `electron-packager` can be found here:
https://www.npmjs.com/package/electron-packager

### Notes

The output app will match your current operational system and arch type.

  * To get an **OSX** app, run it from a **Osx** machine.
  * To get an **Linux 32bit** app, run it from a **32bit Linux** machine.
  * To get an **Linux 64bit** app, run it from a **64bit Linux** machine.
  * To get an **Windows 32bit** app, run it from a **32bit Windows** machine.
  * To get an **Windows 64bit** app, run it from a **64bit Windows** machine.

Due to NodeJS native bindings of such libraries such as Fibers -- *which are
mandatory for Meteor*, you'll need to have your Meteor app fully working on the
desired platform before installing this plugin and packaging your app.

So, at this time, you cannot package your app in a cross-platform fashion from
one single OS.

Perhaps you can live with it? :)

> **DO NOT** use options to output for multiple arch/platforms at once, such as
`--arch=all`. It won't work, Electrify can bundle Meteor apps only for the
platform you're running on.


## Options

1. `-i, --input` - Meteor app folder, default is current directory (`process.cwd()`).
1. `-o, --output` - Sets output folder for your packaged app, default is
`/your/meteor/app/.dist`
1. `-s, --settings` Sets path for Meteor
[settings](http://docs.meteor.com/#/full/meteor_settings) file, this will be
available inside your Meteor code both in development and after being packaged.

## Structure

You'll notice a new folder called `.electrify` in your meteor app dir, its
structure will be like this:

````
/your/meteor/app
├── .electrify
│   ├── .gitignore
│   ├── electrify.json
│   ├── index.js
│   └── package.json
├── .meteor
└── ...
````

This is a pure Electron project, so you can use the whole Electron API from JS
files in this folder. Also, you can install electron dependencies and store them
in the `package.json` file. Note that the `electrify` package is itself a
dependency.

See this folder as the `desktop layer` for your Meteor app. Remember to check
out the `index.js` file, it constains the electrify start/stop usage.

The `electrify.json` file will hold specific preferences for Electrify, such as
plugins and so on. It's still a WIP, but you can get around it.

### Config (`electrify.json`)

For now there's only one option here: `preserve_db`.

Set it to true to preserve database between installs. It works by saving the
mongo data dir inside user's data folder, instead of being self contained within
the app folder (which gets deleted when new version is installed).

# Customizing

Let's see how one would be able to do a simple SplashScreen:

````javascript
var app       = require('app');
var browser   = require('browser-window');
var electrify = require('electrify')(__dirname);

var window = null;
var splash = null; // splash variable

app.on('ready', function() {
  splash = new browser({ // starts splash window
    // >>> your configs here
  });
  splash.loadUrl('./splash.html'); // create the ".electrify/splash.html" file
  
  // then move along and start electrify
  electrify.start(function(meteor_root_url) {

    // before opening a new window with your app, destroy the splash window
    splash.close(); // >>> or .destroy(), check what works for you


    // from here on, well, nothing changes..


    window = new browser({
      width: 1200, height: 900,
      'node-integration': false // node integration must to be off
    });
    window.loadUrl(meteor_root_url);
  });
});


// ....

````

## Meteor x Electron integration

You can seamlessly call Electron methods from your Meteor's client/server code.

Define your Electron methods inside the `.electrify` folder:

````javascript
// `.electrify/index.js` file
electrify.methods({
  'hello.world': function(firstname, lastname, done) {
    // do things with electron api, and then call the `done` callback
    // as ~> done(err, res);
    done(null, 'Hello '+ firstname +' '+ lastname +'!');
   }
});
````

Then, in your Meteor code (client and server), you can call this method like:

````javascript
// Electrify.call(method_name, args, done_callback);
Electrify.call('hello.world', ['anderson', 'arboleya'], function(err, msg) {
  console.log(msg); // Hello anderson arboleya!
});
````

> **IMPORTANT**
> 
> You can only call methods after the connection is made between Meteor and
> Electron, to make sure it's ready you can wrap your code in a startup block:
> 
> ````javascript
> Electrify.startup(function(){
>   Electrify.call(...);
> });
> ````

## Upgrading

When upgrading to newer versions, it's **important** to know that:

### ~> templates

Once these files exists on disk, they *will not* be overwritten.
 * `.electrify/index.js`
 * `.electrify/package.json`
 * `.electrify/electrify.json`
 * `.electrify/.gitignore.json`

### ~> api

As these files above is never overwritten, in case of any API change that needs
adjustments, these will have to be made manually.

### ~> version matching

Always keep the same electrify version in your Meteor, and inside the
`.electrify` folder, *as per specified in `.electrify/package.json` file*.

## Questions?

Do not open issues, use the chat channel instead.

[![Join the chat at https://gitter.im/arboleya/electrify](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/arboleya/electrify)

## Problems?

This is very young and active software, so make sure your are always up to date
before opening an issue. Follow the released fixes through the
[HISTORY.md](HISTORY.md) file.

If you find any problem, please open a meaningful issue describing in detail how
to reproduce the problem, which platform/os/arch type you're using, as well as
the version of Meteor and Electrify, and any other info you may find usefull.

## License

The MIT License (MIT)

Copyright (c) 2015 Anderson Arboleya

HISTORY.md
2.1.5 / 2016-03-15
===================
  * upgrading electron to last version

2.1.4 / 2015-12-14
===================
  * fixing windows env initialization
  * properly spawning system npm on windows

2.1.3 / 2015-12-10
===================
  * Fixig Meteor's settings evaluation

2.1.2 / 2015-12-09
===================
  * Using system NPM to perform installs
  * Automatically installing Electrify meteor package
  * Upgrading all dependencies to last versions

2.1.1 / 2015-12-08
===================
  * Fixing broken release

2.1.0 / 2015-12-08
===================
  * Added `preserve_db` option

2.0.2 / 2015-11-09
===================
  * Fixing last release (broken due to a wrong variable inside a conditional)

2.0.1 / 2015-11-09
===================
 * Fixing requirememts for production app (meteor is not required!)
 * Fixing packager options parser (in cli.js)

2.0.0 / 2015-11-03
===================
  * Added integration/communication between Electron and Meteor
  * Plugins architecture re-designed
  * Fixed process termination
  * Fixed Meteor's settings
  * Redesigned core dependencies
  * Core project is now NPM, the Meteor package only provides a communication
  channel between Meteor and Electron, all the rest is done with the NPM package
  * Adding more options to the CLI tool for more flexibility

1.4.0 / 2015-10-08
===================
  * Improving plugins API

1.3.6 / 2015-10-06
===================
  * Unifying the moment when mongod.lock file is removed

1.3.5 / 2015-10-06
===================
  * Properly handling mongodb.lock file, specially in winows where's not
  possible to gracefully shutdown mongo
  * Fixing meteor initialization on windows

1.3.4 / 2015-10-02
===================
  * Fixing log level check

1.3.3 / 2015-10-02
===================
  * Fixing `package.json` version evaluation for .electrify app

1.3.2 / 2015-10-02
===================
  * Adding target to maintain all versions of meteor and npm packages under
  the same rule

1.3.1 / 2015-10-02
===================
  * Fixing SyntaxError in default template (index.js)

1.3.0 / 2015-10-02
===================
  * Revamping project, re-writting everything
  * Starting initial architecture for plugins
  * Adding tests

1.2.2 / 2015-09-14
===================
  * Executing dependencies' (electron, electron-packager, npm) through it's JS
  files (instead of its binaries) for more cross-platform compatibility and
  flexibility (closes [#8](https://github.com/arboleya/electrify/issues/8))

1.2.1 / 2015-09-09
===================
  * Fixing Meteor's setting format

1.2.0 / 2015-09-09
===================
  * Fixing version checkup
  * Adding support for `meteor --settings` from both Meteor and NPM

1.1.3 / 2015-09-08
===================
 * Adding auto self-upgrade routine for `.electrify` local npm module

1.1.2 / 2015-09-07
===================
 * Improving free port evaluation approach
 * Using Meteor's distributed `npm` to minimize permission problems often found
 when using user's `npm`

1.1.1 / 2015-08-25
===================
 * Fixing project's root folder evaluation approach

1.1.0 / 2015-08-11
===================
 * Adding alternative method to package app, using purely NPM.

1.0.4 / 2015-07-15
===================
 * Removing posix library, using a cross-platform approach based on
 `process.env.METEOR_PARENT_PID` for avoiding opening multiple Electorn windows
 on server restart

1.0.3 / 2015-07-15
===================
 * Avoid opening a new Electron window per server restart, during development

1.0.2 / 2015-07-13
===================
 * Checking and fixing corrupted/incomplete `npm install`s

1.0.1 / 2015-07-12
===================
 * Leaving development database behind
  * Due to unknown erros when starting up mongod for existent files db copied
  from meteor, the electrified app will now start with a blank database, so any
  seeds your app has will be done again in the first run of your packaged app

1.0.0 / 2015-07-12
===================
 * Hello World


index.meteor.js
/*******************************************************************************
  BASE REQUIREMENTS
*******************************************************************************/

var is_meteor = 'undefined' != typeof(Meteor);
var require   = (is_meteor ? Npm.require : require);

require('shelljs/global');

var fs    = require('fs');
var path  = require('path');
var os    = require('os');
var spawn = require('child_process').spawn;

var join = path.join;

/*******************************************************************************
  VARIABLES
*******************************************************************************/

// current version
VERSION = '1.2.2';

// system
_PLATFORM = process.platform;
_ARCH     = process.arch;
_OS       = (_PLATFORM === 'darwin' ? 'osx' : _PLATFORM);
_OS_HOME  = (_PLATFORM == 'win32' ? 'USERPROFILE' : 'HOME');
_TMP      = os.tmpdir();

// silent execs
_SILENT_EXECS = {silent: process.env.BUTTER !== 'true'};

// general folders
_HOME        = process.env[_OS_HOME];
_APP_ROOT    = path.resolve(is_meteor ? '../../../../..' : '.');

// elecritifed lots of folders
_ELECTRIFIED          = join(_APP_ROOT             , '.electrify');
_ELECTRIFIED_RELEASE  = join(_ELECTRIFIED          , '.dist');
_ELECTRIFIED_INDEX    = join(_ELECTRIFIED          , 'index.js');
_ELECTRIFIED_PKG      = join(_ELECTRIFIED          , 'package.json');
_ELECTRIFIED_IGNORE   = join(_ELECTRIFIED          , '.gitignore');
_ELECTRIFIED_TEMP     = join(_ELECTRIFIED          , 'temp');
_ELECTRIFIED_APP      = join(_ELECTRIFIED          , 'app');
_ELECTRIFIED_BIN      = join(_ELECTRIFIED          , 'bin');

// electrified mongo
_ELECTRIFIED_DB       = join(_ELECTRIFIED          , 'db');
_ELECTRIFIED_BUNDLE   = join(_ELECTRIFIED_TEMP     , 'bundle');
_ELECTRIFIED_SERVER   = join(_ELECTRIFIED_APP      , 'programs', 'server');

// electrified node_modules
_ELECTRIFIED_MODS     = join(_ELECTRIFIED          , 'node_modules');
_ELECTRIFIED_ELECTRON = join(_ELECTRIFIED_MODS     , 'electron-prebuilt');
_ELECTRIFIED_PACKAGER = join(_ELECTRIFIED_MODS     , 'electron-packager');

_ELECTRON_PKG_JSON    = join(_ELECTRIFIED_ELECTRON , 'package.json');

_ELECTRIFIED_MODS_B   = join(_ELECTRIFIED_MODS     , '.bin');

_ELECTRON             = join(_ELECTRIFIED_ELECTRON   , 'cli.js');
_ELECTRON_PACKAGER    = join(_ELECTRIFIED_PACKAGER   , 'cli.js');

// meteor stuff
if(_PLATFORM === 'win32') {
  
  _METEOR_FOLDER  = path.join(_HOME, 'AppData', 'Local', '.meteor');
  _METEOR_SYMLINK = cat(path.join(_METEOR_FOLDER, 'meteor.bat'));
  _METEOR_SYMLINK = _METEOR_SYMLINK.match(/\\packages.*\\meteor\.bat/)[0];
  _METEOR_BIN     = path.join(_METEOR_FOLDER, _METEOR_SYMLINK);

} else {
  
  _METEOR_FOLDER  = path.join(_HOME, '.meteor');
  _METEOR_SYMLINK = fs.readlinkSync(path.join(_METEOR_FOLDER, 'meteor'));
  _METEOR_BIN     = path.join(_METEOR_FOLDER, _METEOR_SYMLINK);

}

// meteor folders
_METEOR_TOOLS_FOLDER    = _METEOR_BIN.replace(/meteor(\.bat)?$/m, '');
_METEOR_DEV_BUNDLE      = path.join(_METEOR_TOOLS_FOLDER, 'dev_bundle');
_METEOR_SERVER_LIB      = path.join(_METEOR_DEV_BUNDLE, 'server-lib');
_METEOR_SERVER_LIB_MODS = path.join(_METEOR_SERVER_LIB, 'node_modules');

// meteor's mongodb distro
_METEOR_MONGO  = path.join(_METEOR_DEV_BUNDLE, 'mongodb', 'bin', 'mongo');
_METEOR_MONGOD = _METEOR_MONGO + 'd';

// meteor's nodejs + npm distro
_METEOR_NODE = path.join(_METEOR_DEV_BUNDLE, 'bin', 'node');

if(_PLATFORM === 'win32')
  _METEOR_MODS = path.join(_METEOR_DEV_BUNDLE, 'bin', 'node_modules');
else
  _METEOR_MODS = path.join(_METEOR_DEV_BUNDLE, 'lib', 'node_modules');

_METEOR_NPM  = path.join(_METEOR_MODS, 'npm', 'bin', 'npm-cli.js');
_METEOR_LOCAL_DB = path.join(_APP_ROOT, '.meteor', 'local', 'db');

if(is_meteor)
  _METEOR_SETTINGS = Meteor.settings;
else {
  _METEOR_SETTINGS = _NPM_SETTINGS;
}

// bin ref differences
if(_PLATFORM === 'win32') {
  _METEOR_MONGO  += '.exe';
  _METEOR_MONGOD += '.exe';
  _METEOR_NODE   += '.exe';
}

// database
_METEOR_LOCAL_DB_JOURNAL = path.join(_METEOR_LOCAL_DB, 'journal');
_METEOR_LOCAL_DB_FILES   = [
  path.join(_METEOR_LOCAL_DB, 'local.ns'),
  path.join(_METEOR_LOCAL_DB, 'local.0'),
  path.join(_METEOR_LOCAL_DB, 'local.1'),
  path.join(_METEOR_LOCAL_DB, 'meteor.ns'),
  path.join(_METEOR_LOCAL_DB, 'meteor.0'),
  path.join(_METEOR_LOCAL_DB, 'meteor.1')
];

/*******************************************************************************
  INITIALIZATION
*******************************************************************************/

this.__defineGetter__('electrify', release);

if(is_meteor) {
  // in development mode this method is called everytime server is restarted due
  // to some server file change, so we need some trick to avoid opening a new
  // Electron window everytime it happens - see the ppid stuff below
  Meteor.startup(function(){

    // ELECTRON_PRODUCTION variable comes from the `index.js` file, check its
    // initialization section to get wtf is going on here - basically, since the
    // app has bem packaged, this whole method is useless, so we simply abort
    if(process.env.ELECTRON_PRODUCTION) return;
    
    // now we fetch the meteor previous and current PID in order to compare them
    var ppid_filepath = path.join(_TMP, '.electrify-ppid')
    var previous_ppid = read(ppid_filepath);
    var current_ppid  = process.env.METEOR_PARENT_PID;

    // if the current ppid differs from the last, we assume some SIGINT has
    // ocurred on the meteor-tool command line, so it's fine to launch electron
    // window again otherwise we simply assume it's a mere server restart, close
    // our eyes and relax :)
    if(previous_ppid != current_ppid) {

      // we also save the current ppid for next comparison
      write(ppid_filepath, current_ppid);

      // and start everything for development
      development();
    }
  });
}

/*******************************************************************************
  STARTUP MODES
*******************************************************************************/

if(!is_meteor) exports.release = release;

function development(){
  setup_folders();
  copy_templates();
  install_electrified_dependencies();
  launch_electron();
}

function release(){
  setup_folders();
  copy_templates();
  copy_bins();
  // copy_database();
  bundle_meteor();
  install_meteor_dependencies();
  install_electrified_dependencies();
  package_app();
}

/*******************************************************************************
  FOLDERS
*******************************************************************************/

function setup_folders(){
  if(!exists(_ELECTRIFIED)) {
    log('setting up folders');
    mkdir('-p', _ELECTRIFIED);
  }
}

/*******************************************************************************
  TEMPLATES
*******************************************************************************/

function copy_templates() {
  if(!exists(_ELECTRIFIED_INDEX))
    log('copying template files');

  if(!exists(_ELECTRIFIED_PKG))
    write(_ELECTRIFIED_PKG, TEMPLATES.pkg);
  
  if(!exists(_ELECTRIFIED_INDEX))
    write(_ELECTRIFIED_INDEX, TEMPLATES.index);

  if(!exists(_ELECTRIFIED_IGNORE))
    write(_ELECTRIFIED_IGNORE, TEMPLATES.gitignore);
}

/*******************************************************************************
  INSTALL DEPENDENCIES
*******************************************************************************/

function install_electrified_dependencies() {

  // in dev mode, link local copy of electrify module :)
  if(process.env.BUTTER === 'true') {

    var source  = path.join(_APP_ROOT, '..', 'electrify');
    var symlink =  path.join(_ELECTRIFIED_MODS, 'electrify');
    
    if(!exists(_ELECTRIFIED_MODS))
      mkdir('-p', _ELECTRIFIED_MODS);


    // when in dev mode, link electrify npm package right into node_modules
    rm('-rf', symlink);
    ln('-s', source, symlink);
  }

  // checks for a possible previous corrupted install of modules, this can
  // happen if the network fails during the `npm_install` command or other
  // similar case that precociously interupt it, preventing it from downloading
  // and building electron-prebuilt or electron-packager modules, gracefully

  // first we check if `node_modules` folder exists
  if(exists(_ELECTRIFIED_MODS)) {

    // then we check if some of the needed binaries is not present
    var binaries_missing = !exists(_ELECTRON) || !exists(_ELECTRON_PACKAGER);

    // or local npm version is out of date
    var package_json = require(_ELECTRIFIED_PKG)
    var outdated = package_json.devDependencies.electrify != VERSION;

    if(binaries_missing || outdated) {
      // in case something is missing, it means something went wrong during the
      // `npm_install` command, preventing it from finishing downloading or
      // building everything properly

      // so we remove the `node_modules` folder to erase any trace of these
      // corrupted/incomplete install process before procceding and installing
      // everything again
      rm('-rf', _ELECTRIFIED_MODS);

      // update local copy of npm electify module
      if(outdated) {
        log('upgrading local `.electrify` npm module to `v'+ VERSION +'`');
        package_json.devDependencies.electrify = VERSION;
        write(_ELECTRIFIED_PKG, JSON.stringify(package_json, null, 2));
      }
    } else {
      return;
    }
  }

  log('installing electrified dependencies');
  var cmd = ['cd', _ELECTRIFIED, '&&', _METEOR_NODE, _METEOR_NPM, 'install'];
  exec(cmd.join(' '), _SILENT_EXECS);
}

/*******************************************************************************
  ELECTRON LAUNCHER
*******************************************************************************/

function launch_electron() {
  log('launching electron');
  spawn(_METEOR_NODE, [_ELECTRON, _ELECTRIFIED], {stdio: 'inherit'});
}

/*******************************************************************************
  INJECTIONS
*******************************************************************************/

function copy_bins() {
  if(!exists(_ELECTRIFIED_BIN)){
    mkdir('-p', _ELECTRIFIED_BIN);
    log('copying mongo and node binaries');
    cp(_METEOR_MONGO, _ELECTRIFIED_BIN);
    cp(_METEOR_MONGOD, _ELECTRIFIED_BIN);
    cp(_METEOR_NODE, _ELECTRIFIED_BIN);
  }
}

function copy_database() {
  log('copying database');
  rm('-rf', _ELECTRIFIED_DB);
  mkdir('-p', _ELECTRIFIED_DB);
  cp('-r', _METEOR_LOCAL_DB_JOURNAL, _ELECTRIFIED_DB);
  _METEOR_LOCAL_DB_FILES.forEach(function(filepath){
    if(exists(filepath))
      cp('-r', filepath, _ELECTRIFIED_DB);
  });
}

/*******************************************************************************
  BUNDLE METEOR APP
*******************************************************************************/

function bundle_meteor() {
  log('bundling meteor');

  exec([
    'cd ' + _APP_ROOT,
    'meteor build ' + _ELECTRIFIED_TEMP + ' --server null --directory'
  ].join(' && '), _SILENT_EXECS);

  var settings_path = path.join(_ELECTRIFIED_BUNDLE, 'settings.json');
  write(settings_path, JSON.stringify(_METEOR_SETTINGS));

  rm('-rf', _ELECTRIFIED_APP);
  mv(_ELECTRIFIED_BUNDLE, _ELECTRIFIED_APP);
  rm('-rf', _ELECTRIFIED_TEMP);
}

function install_meteor_dependencies() {
  log('installing meteor dependencies');
  // instead of `npm install` :
  // 
  //     exec('cd ' + _ELCTRIFIED_SERVER + ' && npm install', _SILENT_EXECS);
  // 
  // which would imply in another bugs around node-fibers native re-build with
  // node-gyp, we just copy the whole `node_modules` folder that is officially
  // distributed with meteor, its 'ready to go and doesn't need to be rebuilt
  cp('-r', _METEOR_SERVER_LIB_MODS, _ELECTRIFIED_SERVER);
}

/*******************************************************************************
  PACKAGE APP
*******************************************************************************/

function package_app() {
  var name    = require(_ELECTRIFIED_PKG).name;
  var version = require(_ELECTRON_PKG_JSON).version;

  var command = [
    _METEOR_NODE + ' ',
    _ELECTRON_PACKAGER +' '+ _ELECTRIFIED + ' ' + name,
    '--out=' + _ELECTRIFIED_RELEASE,
    '--arch=' + _ARCH,
    '--platform=' + _PLATFORM,
    '--version='+ version
  ].join(' ');

  log(
    'packaging app for platform '+ _PLATFORM +
    ' '+ _ARCH + ' using electron v' + version
  );

  rm('-rf', _ELECTRIFIED_RELEASE);
  mkdir('-p', _ELECTRIFIED_RELEASE);

  exec(command, _SILENT_EXECS);

  log('wrote new app to ', _ELECTRIFIED_RELEASE);
}


/*******************************************************************************
  TEMPLATES
*******************************************************************************/

TEMPLATES = {
  index: [
    "var app       = require('app');",
    "var browser   = require('browser-window');",
    "var electrify = require('electrify');",
    "",
    "",
    "var window    = null;",
    "",
    "app.on('ready', function() {",
    "",
    "  window = new browser({",
    "    width: 1200,",
    "    height: 900,",
    "    'node-integration': false",
    "  });",
    "  ",
    "  electrify.boot(function() {",
    "    window.loadUrl(electrify.meteor_url);",
    "  });",
    "",
    "});",
    "",
    "",
    "app.on('will-quit', function(event) {",
    "  electrify.shutdown(app, event);",
    "});",
    "",
    "app.on('window-all-closed', function() {",
    "  app.quit();",
    "});"
  ].join('\r\n'),

  pkg: [
  '{',
  '  "name": "Electrified",',
  '  "main": "index.js",',
  '  "devDependencies": {',
  '    "electron-prebuilt": "^0.29.2",',
  '    "electron-packager": "^5.0.1",',
  '    "electrify": "'+ VERSION +'"',
  '  }',
  '}'
  ].join('\r\n'),

  gitignore: [
    '.DS_Store',
    '.dist',
    'app',
    'bin',
    'db',
    'node_modules'
  ].join('\r\n')
};

/*******************************************************************************
  SIMPLE FILE READ / WRITE / EXISTS helpers
*******************************************************************************/

function read(filepath, encoding) {
  if(exists(filepath))
    return fs.readFileSync(filepath, encoding || 'utf8');
}

function write(filepath, contents) {
  fs.writeFileSync(filepath, contents);
}

function exists(filepath) {
  return fs.existsSync(filepath);
}

/*******************************************************************************
  DEBUG
*******************************************************************************/

function log(){
  var args = Array.prototype.slice.call(arguments);
  console.log.apply(null, ['electrify: '].concat(args));
}